packages/devtools_app/lib/src/debugger/evaluate.dart

```diff
@@ -24,17 +29,142 @@ class ExpressionEvalField extends StatefulWidget {
+        searchingValue,
+        selection,
+        handleFields: isField,
+      );
+
+      // Only show pop-up if there's a real variable name or field.
+      if (parts.activeWord.isEmpty && !parts.isField) return;
+
+      final matches = await _matchesFor(parts);
+
+      final normalizedMatches = matches.toSet().toList()..sort();
+      _autoCompleteController.searchAutoComplete.value =
+          normalizedMatches.sublist(
+        0,
+        min(defaultTopMatchesLimit, normalizedMatches.length),
+      );
+    } else {
+      _autoCompleteController.closeAutoCompleteOverlay();
+    }
+  }
+
+  Future<List<String>> _matchesFor(EditingParts parts) async {
+    final result = <String>{};
+    if (!parts.isField) {
+      for (var variable in widget.controller.variables.value) {
+        if (variable.boundVar.name.startsWith(parts.activeWord)) {
+          result.add(variable.boundVar.name);
+        }
+      }
+    } else {
+      try {
+        var left = parts.leftSide.split(' ').last;
+        // Removing trailing `.`.
+        left = left.substring(0, left.length - 1);
+
+        // Response is either a ErrorRef, InstanceRef, or Sentinel.
+        final response = await widget.controller.evalAtCurrentFrame(left);
+
+        // Display the response to the user.
+        if (response is InstanceRef) {
+          final Instance instance = await widget.controller.getObject(response);
+          result.addAll(await _autoCompleteProperties(instance.classRef));
+          // TODO(grouma) - This shouldn't be necessary but package:dwds does
+          // not properly provide superclass information.
+          result.addAll(instance.fields.map((field) => field.decl.name));
+          result.removeWhere((prop) => !prop.startsWith(parts.activeWord));
+        }
+      } catch (_) {}
+    }
+    return result.toList();
+  }
+
+  Future<List<String>> _autoCompleteProperties(ClassRef classRef) async {
+    final result = <String>[];
+    if (classRef != null) {
+      final Class clazz = await widget.controller.getObject(classRef);
+      result.addAll(clazz.fields.map((field) => field.name));
+      result.addAll(clazz.functions
+          .where((funcRef) => _validFunction(funcRef, clazz))
+          // The VM shows setters as `<member>=`.
+          .map((funcRef) => funcRef.name.replaceAll('=', '')));
+      result.addAll(await _autoCompleteProperties(clazz.superClass));
+      result.removeWhere((member) => !_isAccessible(member, clazz));
+    }
+    return result;
+  }
+
+  bool _validFunction(FuncRef funcRef, Class clazz) {
+    return !funcRef.isStatic &&
+        !_isContructor(funcRef, clazz) &&
+        funcRef.name != '==';
+  }
+
+  bool _isContructor(FuncRef funcRef, Class clazz) =>
+      funcRef.name == clazz.name || funcRef.name.startsWith('${clazz.name}.');
+
+  bool _isAccessible(String member, Class clazz) {
+    final frame = widget.controller.selectedStackFrame.value?.frame ??
```

Comment:
why is the `??` needed. Can we just always have the first frame be selected? Seems like that is the UI state in the debugger anyway.

Reply:
This is a bit of copied code from the evaluation logic. I'm not exactly sure why it's required but the added defensiveness seemed important.

Reply:
I would add it to the controller. Maybe "currentStackFrame".